[WIP] Reconstruction constructor in CoordSys3D

szymag opened the pull request:
This PR introduces to constructor transformation equation. 
These equations will be obtain from `rotation_matrix`, `location` (for translations), and for changing type of coordinate system. After obtaining these equations, method for transformation composition will be used.

We need to deal with two situation. When we want to connect our coordinate system (parent coordinate system) and when we want to just set transformation equations.

szymag committed:
Add structure for transformation equations in constructor.
Add rotation_equations.

szymag committed:
Merge branch 'master' into reconstruction_constructor

szymag committed:
Add transformation equations.
Set rotation_equations to None for parent None and rotation_matrix None

Upabjojr committed:
Update coordsysrect.py

Upabjojr commented on sympy/vector/coordsysrect.py:
```diff
@@ -395,19 +421,20 @@ def _inverse_rotation_matrix(self):
+        trans_eq2 = matrix[3]*variables[0] + matrix[4]*variables[1] + matrix[5]*variables[2]
+        trans_eq3 = matrix[6]*variables[0] + matrix[7]*variables[1] + matrix[8]*variables[2]
 
         return trans_eq1, trans_eq2, trans_eq3
```
if we operate with `lambda`, it would be easier.

Upabjojr commented on sympy/vector/coordsysrect.py:
```diff
@@ -98,6 +98,33 @@ def __new__(cls, name, location=None, rotation_matrix=None,
             location = Vector.zero
             origin = Point(name + '.origin')
 
```
What to do here:

- check that if `ŧransformation` is `None`.
- if it is None, computer `transformation` from the rotation and traslation. Make identity if they are also None.
- if it is not None, check that rotation and translation are None
- transform `transformation` to a `Lambda`.

szymag replied:
I'm not sure why do that.
You assume that we have to use `transformation` or (`rotation` and `translation`), so `transformation` should have information about every three transformations.
I think that it could be easier sometimes to split transformation into steps. For example for `transformation` set `cylindrical` and for location set some `vector`.

Upabjojr replied:
We do it the general way, no need to complicate the code with special cases. Translations are subsets of transformations. In the special cases of rotation and translations, we could store some markers to remind of the special case (and remember the steps).

Upabjojr commented:
I suggest to return `lambda` objects from the methods, so you can compose the transformations by simply calculating `f(*g(x))`.

szymag committed:
Update _rotation_trans_equations, _translation_trans_equations.

szymag committed:
Update constructor.

szymag commented:
> I suggest to return lambda objects from the methods, so you can compose the transformations by simply calculating f(*g(x)).

I used your idea in constructor, when `rotation_equations` and `translation_equations` are obtained.

szymag committed:
Check orthogonality in constructor.

Upabjojr commented on sympy/vector/coordsysrect.py:
```diff
@@ -109,12 +156,11 @@ def __new__(cls, name, location=None, rotation_matrix=None,
         if parent is not None:
             obj = super(CoordSys3D, cls).__new__(
-                cls, Symbol(name), location, parent_orient, parent)
+                cls, Symbol(name), transformation, parent)
```
`transformation` need to be defined.

Upabjojr replied:
`transformation` has to be a SymPy type, if we are going to use a lambda, we need to use SymPy's `Lambda` (with capital L).

Upabjojr commented on sympy/vector/coordsysrect.py:
```diff
@@ -98,6 +98,53 @@ def __new__(cls, name, location=None, rotation_matrix=None,
+            if parent is None:
+                transformation_equations = None
+            else:
+                transformation_equations = None
```
The aim is to define `transformation`, we may not need `transformation_equations`.

Upabjojr commented on sympy/vector/coordsysrect.py:
```diff
@@ -98,6 +98,53 @@ def __new__(cls, name, location=None, rotation_matrix=None,
+                raise ValueError
+
+            if parent is None:
+                transformation_equations = None
```
Why `None`? We have `transformation != None` in this case. We should just check that `transformation` is correct (and transform it to a SymPy type).

szymag replied:
I didn't implement this part yet.
I just put `None` to create some structure for case when `transformation` is not `None`.

szymag replied:
I think that we need to think first, how `transformation` should look like. Should we just use `_connect_to_standard_cartesian`?

Upabjojr replied:
That code can be OK. I would prefer that `transformation` be finally transformed into a `Lambda` object, it makes more sense with the hashing.

Upabjojr commented on sympy/vector/coordsysrect.py:
```diff
@@ -174,7 +220,7 @@ def __new__(cls, name, location=None, rotation_matrix=None,
         obj._parent_rotation_matrix = parent_orient
         obj._origin = origin
-
+        obj.transformation_equations = transformation_equations
```
Store `transformation`, not `transformation_equations` (they are the same thing in the end).

szymag committed:
Update code appling Upabjojr suggestion.

szymag committed:
Adapt methods for transformation.

Upabjojr commented on sympy/vector/coordsysrect.py:
```diff
@@ -98,6 +213,65 @@ def __new__(cls, name, location=None, rotation_matrix=None,
+                lame_coefficients = connect_to[1]
+                inverse_transformation = connect_to[2]
+            else:
+                transformation = None
```
This is definitely wrong.

szymag replied:
I didn't cover this scenario. When you accept the case when `parent` is `None` I will update this part. 
I wrote about it on gitter

Upabjojr replied:
Let's remove the `parent` case. Sorry, didn't read gitter.

Upabjojr commented on sympy/vector/coordsysrect.py:
```diff
@@ -98,6 +213,65 @@ def __new__(cls, name, location=None, rotation_matrix=None,
+                # See lines 80:99. Location is somehow treated only when parent is put.
+                # We are not able to create independent vector, so we can't assign
+                # anything to location.
+                translation_equations = lambda x, y, z: (x, y, z)
```
I believe we should treat the parent as being a default standard coordinate system if it is not specified. Otherwise what sense does it make to specify the transformation?

Upabjojr commented on sympy/vector/coordsysrect.py:
```diff
@@ -98,6 +213,65 @@ def __new__(cls, name, location=None, rotation_matrix=None,
+                x, y, z = symbols('x y z')
+                transformation = rotation_equations(*translation_equations(x, y, z))
+                lame_coefficients = _set_lame_coefficient_mapping('cartesian', (x, y, z))
+                inverse_transformation = None
```
Well, this could be specified. You just have to immagine that there is some default coordinate system used as a reference if `parent == None`.

Upabjojr commented on sympy/vector/coordsysrect.py:
```diff
@@ -98,6 +213,65 @@ def __new__(cls, name, location=None, rotation_matrix=None,
+                lame_coefficients = _set_lame_coefficient_mapping('cartesian', (x, y, z))
+                inverse_transformation = None
+
+                if not CoordSys3D._check_orthogonality(transformation, (x, y, z)):
```
If there are only rotations and translations, we don't need this check.

Upabjojr commented on sympy/vector/coordsysrect.py:
```diff
@@ -98,6 +213,65 @@ def __new__(cls, name, location=None, rotation_matrix=None,
             origin = Point(name + '.origin')
 
+        if transformation is None:
+            if parent is None:
```
This `if` loop is not very useful. The case where `parent == None` and `parent != None` could be handled by the same code.

szymag replied:
It could be, but the one difference between these two cases is with `variables`. When we have `parent` we can construct transformation equations using base scalars from `parent` system. Otherwise we can't.

Upabjojr replied:
We are using lambdas in order to avoid this problem.

szymag replied:
OK, I see. I will simplify this

szymag replied:
Another thing is `translation`. Without `parent` it doesn't have sense, so we need to distinguish two cases. But I think we can still simplify the code a bit.

Upabjojr replied:
Why doesn't it make sense? Anyway, the code in the master supports location even without parent, so we need to keep the compatibility.

Upabjojr commented on sympy/vector/coordsysrect.py:
```diff
@@ -98,6 +213,65 @@ def __new__(cls, name, location=None, rotation_matrix=None,
+                lame_coefficients = _set_lame_coefficient_mapping('cartesian', parent.base_scalars())
+                inverse_transformation = _set_inv_trans_equations('cartesian', parent.base_scalars())
+
+                if not CoordSys3D._check_orthogonality(transformation, parent.base_scalars()):
```
useless check, it will always pass (for rotations and translations).

szymag replied:
OK

Upabjojr commented on sympy/vector/coordsysrect.py:
```diff
@@ -98,6 +213,65 @@ def __new__(cls, name, location=None, rotation_matrix=None,
+            if location is not None and rotation_matrix is not None:
+                raise ValueError
+
+            if parent is None:
```
I would remove this `if`.

Upabjojr commented on sympy/vector/coordsysrect.py:
```diff
@@ -196,45 +355,53 @@ def _connect_to_standard_cartesian(self, curv_coord_type, inverse=True):
+            else:
+                inverse_transformation_equations = None
+
+        elif isinstance(curv_coord_type, (tuple, list, Tuple, collections.Callable)):
```
`Callable` should be a separate case.

Upabjojr committed:
Update coordsysrect.py

Upabjojr committed:
Update coordsysrect.py

szymag committed:
Update CoordSys3D.

Almost finished constructor.

Lame_coefficinets, transformation, inverse_transfromation are Lambda,
so they can easily depend on desired variables.

Creating base scalars with custom str has the same form as base scalars
created in standard name: class_name + '.' + variable_names.

Upabjojr commented on sympy/vector/coordsysrect.py:
```diff
@@ -225,28 +192,29 @@ def __new__(cls, name, transformation=None, parent=None, location=None,
-                    rotation_lambda = lambda x, y, z: (x, y, z)
+            if rotation_matrix is not None:
+                rotation_lambda = \
+                    lambda x, y, z: parent._rotation_trans_equations(parent_orient, (x, y, z))
```
with `lambda` no need for `parent` here.

Upabjojr commented on sympy/vector/coordsysrect.py:
```diff
@@ -310,9 +278,12 @@ def __new__(cls, name, transformation=None, parent=None, location=None,
-        obj._inverse_transformation = inverse_transformation
+        obj._h1, obj._h2, obj._h3 = lame_coefficients(obj._x, obj._y, obj._z)
+        if parent is not None:
+            obj._transformation = transformation(parent._x, parent._y, parent._z)
```
I think `transformation` should remain a `Lambda`.

Upabjojr commented on sympy/vector/coordsysrect.py:
```diff
@@ -310,9 +278,12 @@ def __new__(cls, name, transformation=None, parent=None, location=None,
+            obj._transformation = transformation(parent._x, parent._y, parent._z)
+        else:
+            obj._transformation = transformation(obj._x, obj._y, obj._z)
+        obj._inverse_transformation = inverse_transformation(obj._x, obj._y, obj._z)
```
also this one, I think it's easier if it remains in the `lambda` form.

Upabjojr commented on sympy/vector/coordsysrect.py:
```diff
@@ -346,19 +317,16 @@ def _connect_to_standard_cartesian(cls, curv_coord_type, inverse=True):
-            if variables is None:
-                variables = _default_variables(curv_coord_type)
-
             transformation_equations = _set_transformation_equations_mapping(curv_coord_type)
```
`transformation_equations` is a `lambda`, we need a SymPy `Lambda` (capital L) for hashing and internal stuff of SymPy.

Upabjojr commented on sympy/vector/coordsysrect.py:
```diff
@@ -367,35 +335,33 @@ def _connect_to_standard_cartesian(cls, curv_coord_type, inverse=True):
+                raise ValueError("The transformation equation does not "
+                                 "create orthogonal coordinate system")
+
+            lame_coefficients = cls._calculate_lame_coefficients(transformation_equations(*variables), variables)
```
do we really need to pass `variables` here? I believe a single `lambda` is enough.

szymag replied:
It was the most difficult part for me and I didn't find clear solution with `lambda` but I can try once again.

Upabjojr replied:
in `_calculate_lame_coefficients` you can use some `Dummy` variables.

szymag replied:
OK, I will try.

szymag committed:
Apply comments.

Remove parameter variables from _calculate_lame_coefficinets,
_calculate_inv_trans_equations, _check_orthogonality

All transformation in constructor are Lambda object.

szymag committed:
Pass tests from test_coordsysret.py.

Inverse transformation is calculated only when user needs it.
when transformation is choosen from pre-defined ones, inverse is
also set.

szymag committed:
Reduce overhead comming from creating transfotmation.

Remove dependency of parameter variables from every method.
Change form of setting transformation. Variable names should be put
together with equations.

Remove simplify in _calculate_lame_coefficients for a while.
Useage of simplify produce errors which isn't resolve yet.

Upabjojr committed:
Huge refactory to CoordSys3D constructor

szymag commented on sympy/vector/coordsysrect.py:
```diff
@@ -98,6 +119,27 @@ def __new__(cls, name, location=None, rotation_matrix=None,
+        if transformation is None:
+            transformation = Tuple(rotation_matrix, location)
+
+        if isinstance(transformation, Tuple):
```
Why that? 
In this case, we are not able to set `transformation` as equations with variables. Previously, we could do that. For example:
```python
a = CoordSys3D('a', transformation=Tuple(((x*cos(y), x*sin(y), z)), (x, y, z))) 
```
But now you assumed that in following manner we're providing rotation and translation transformations. 
Look at the code in `_connect_to_cartesian` method, lines `363:367`

Upabjojr replied:
I thought a bit over the weekend, if we pass a rotation and/or translation, it needs to be stored in `transformation` and passed to the superclass (SymPy's hashing requirements). If we transform the rotation and/or translation into a `Lambda`, the code will slow down a lot, because when the class is reconstructed all orthogonality checks and computations are performed (at least, I think that's the problem).

We can still pass the transformation equations by wrapping them in a `Lambda`.

Upabjojr replied:
The main problem is that when calling the superclass in the constructor, we need a fixed argument order. The same argument order must return the same object if called on the constructor. This requirement is pretty restrictive on the way we may write the constructor, so we need to choose a precise way to pack the information without losing speed.

szymag committed:
Support for custom transformation equations.

szymag committed:
Fix base vectors, base scalars name printing.

Simplification in constructor. Remove unnescessary if condition.

Fix tests, doctests.

Upabjojr committed:
Fixes to gradient/divergence/curl

Upabjojr committed:
inverse transformation for rotation/translation

Upabjojr commented on sympy/vector/tests/test_coordsysrect.py:
```diff
@@ -299,85 +299,60 @@ def test_evalf():
-    raises(ValueError, lambda: a._connect_to_standard_cartesian((x, y, z)))
-
-
-def test_check_orthogonality():
```
why has this test disappeared?

szymag replied:
I must did it earlier, but I shouldn't. I will do it

Upabjojr commented on sympy/vector/tests/test_coordsysrect.py:
```diff
@@ -299,85 +299,60 @@ def test_evalf():
+    assert a.transformation_to_parent() == (a.x * cos(a.y), a.x * sin(a.y), a.z)
     assert simplify(a.lame_coefficients()) == (1, sqrt(a.x**2), 1)
 
-    raises(ValueError, lambda: a._connect_to_standard_cartesian((x, y, z)))
```
we can test invalid parameters to the constructor

Upabjojr commented on sympy/vector/tests/test_printing.py:
```diff
@@ -27,20 +29,20 @@ def upretty(expr):
-v.append(a*N.i - b*N.j)
-v.append((a**2 + N.x)*N.i + N.k)
-v.append((a**2 + b)*N.i + 3*(C.y - c)*N.k)
+v.append(a * N.i)
```
These changes should be reverted.

szymag replied:
I don't know why, but some expression change their order like was `c - C.y` and know `C.y - c`

Upabjojr commented on sympy/vector/tests/test_printing.py:
```diff
@@ -50,15 +52,15 @@ def upretty(expr):
 """)
 pretty_v_11 = u(
-"""\
+    """\
```
revert

szymag replied:
I didn't notice that. I will revert

Upabjojr commented on sympy/vector/tests/test_printing.py:
```diff
@@ -68,26 +70,26 @@ def upretty(expr):
 for x in v:
     d.append(x | N.k)
-s = 3*N.x**2*C.y
+s = 3 * N.x ** 2 * C.y
```
did you use some auto-formatter?

szymag replied:
I think so, by mistake

szymag committed:
Revert unwanted changes

szymag committed:
Revert unwanted changes

Upabjojr commented on sympy/vector/tests/test_coordsysrect.py:
```diff
@@ -299,84 +299,74 @@ def test_evalf():
-    assert a._check_orthogonality() is True
+    x, y, z = symbols('x y z')
+    u,v = symbols('u, v')
+    a = CoordSys3D('a', transformation=((x*sin(y)*cos(z), x*sin(y)*sin(z), x*cos(y)), (x, y, z)))
```
shouldn't the tuple `(x, y, z)` come before the expression?

That's why I suggest to simply use: `lambda x, y, z: (x*sin(y)*cos(z), x*sin(y)*sin(z), x*cos(y))`

szymag replied:
OK

szymag replied:
I will add also support for `lambda`.

Upabjojr commented on sympy/vector/tests/test_coordsysrect.py:
```diff
@@ -299,84 +299,74 @@ def test_evalf():
+    a = CoordSys3D('a', transformation=(((x*cos(y), x*sin(y), z)), (x, y, z)))
+    assert a.transformation_to_parent() == (a.x * cos(a.y), a.x * sin(a.y), a.z)
     assert simplify(a.lame_coefficients()) == (1, sqrt(a.x**2), 1)
-    a._connect_to_standard_cartesian(((x, y, z), (x * sin(y) * cos(z), x * sin(y) * sin(z), x * cos(y))), inverse=False)
```
why is the block from here on still missing?

szymag replied:
We do not support now these type of `transformation`.
Here we mainly tested `_connect_to_standard_cartesian` which now is removed

Upabjojr replied:
Just pass a lambda to `CoordSys3D` constructor. You added support for this tuples.

szymag committed:
Merge branch 'reconstruction_constructor' of https://github.com/szymag/sympy into reconstruction_constructor

szymag committed:
Merge branch 'master' into reconstruction_constructor

szymag committed:
Merge branch 'master' into reconstruction_constructor

szymag committed:
Merge branch 'reconstruction_constructor' of https://github.com/szymag/sympy into reconstruction_constructor

szymag committed:
Revert order in transformation.
Variables now are first in tuple.

Add support for python's lambda.

Add test for wrong transformation argument in constructor.

Fix indent inconstructor.

Upabjojr commented on sympy/vector/coordsysrect.py:
```diff
@@ -62,18 +64,43 @@ def __new__(cls, name, location=None, rotation_matrix=None,
                 raise TypeError("rotation_matrix should be an Immutable" +
                                 "Matrix instance")
-            parent_orient = rotation_matrix
+                rotation_matrix = rotation_matrix.as_immutable()
```
This should be de-indented

szymag replied:
OK

Upabjojr committed:
fixes to directional_derivative, tests restored

szymag committed:
Fix printing error.
